File: 2024/06_guard_gallivant.py
import numpy as np
from torch.linalg import solve

grid = """....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...""".splitlines()
grid = open('data/06.dat').read().splitlines()

grid = [list(row) for row in grid]
starting_position = next((i, j) for i in range(len(grid)) for j in range(len(grid[0])) if grid[i][j] == '^')
obstacles = set((i, j) for i in range(len(grid)) for j in range(len(grid[0])) if grid[i][j] == '#')

directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

def turn(directions):
    directions = directions[1:] + [directions[0]]
    return directions

def solve_one(starting_position, directions, obstacles):
    position = starting_position
    visited = {(starting_position, directions[0])}
    while True:
        x, y = position
        dx, dy = directions[0]
        next_pos = (x + dx, y + dy)

        if next_pos in obstacles:
            directions = turn(directions)
        elif 0 <= next_pos[0] < len(grid) and 0 <= next_pos[1] < len(grid[0]):
            position = next_pos

            if (position, directions[0]) not in visited:
                visited.add((position, directions[0]))
            else:
                return visited, True

        else:
            return visited, False

visited, is_loop = solve_one(starting_position, directions, obstacles)
num_visited_positions = len(set(pos for pos, _ in visited))

print("Part 1: ", num_visited_positions, is_loop)

# Part 2
ctr = 0
for i in range(len(grid)):
    for j in range(len(grid[0])):
        if (i, j) not in obstacles:
            new_obstacles = obstacles | {(i, j)}
            _, is_loop = solve_one(starting_position, directions, new_obstacles)
            ctr += int(is_loop)

print("Part 2: ", ctr)

File: 2024/06_guard_gallivant.jl
example = split("....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...", '\n')


global next_direction = Dict(
    (-1, 0)=>(0, 1), (0, 1)=>(1, 0), (1, 0)=>(0, -1), (0, -1)=>(-1, 0)
)


function parse_input(inp)
    nrows, ncols = length(inp), length(inp[1])
    
    boxes, guard = Set{Tuple{Int, Int}}(), (0, 0)
    for row in range(1, nrows)
        for col in range(1, ncols)
            if inp[row][col] == '#'
                push!(boxes, (row, col))
            elseif inp[row][col] == '^'
                guard = (row, col)
            end
        end
    end
    
    return boxes, guard, nrows, ncols
end


function try_move(boxes, pos, dir)::Tuple{Tuple{Int, Int}, Tuple{Int, Int}}
    #=  This function moves the robot as long as it does not hit a box.
        It returns the new position and does not care whether that position
        is out of bounds. The calling code has to check for that.
        =#
    new_x, new_y = (x + y for (x, y) in zip(pos, dir))

    # Two cases: Either we can move, or we turn right:
    if (new_x, new_y) in boxes
        return pos, next_direction[dir]
    else
        return (new_x, new_y), dir
    end
end


function move(boxes, guard, nrows, ncols)::Int
    # Move the robot until it goes out of bounds OR begins to loop.
    visited = Set{Tuple{Int, Int, Int, Int}}()
    x, y = guard
    dx, dy = (-1, 0)

    while 1 ≤ x ≤ nrows && 1 ≤ y ≤ ncols
        push!(visited, (x, y, dx, dy))
        (x, y), (dx, dy) = try_move(boxes, (x, y), (dx, dy))

        # Check if we are in a loop. If so, return a special value of 0 to indicate that we got an infinite loop.
        if (x, y, dx, dy) in visited
            return 0
        end
    end

    distinct_positions = Set([(x, y) for (x, y, dx, dy) in visited])

    return length(distinct_positions)
end


function check_for_loops(boxes, guard, nrows, ncols)::Int
    return sum([1 for x in range(1, nrows), y in range(1, ncols) if move([boxes..., (x, y)], guard, nrows, ncols) == 0 ])
end


@time begin
    boxes, guard, nrows, ncols = parse_input(readlines("data/06.dat"))
    println("Part 1: ", move(boxes, guard, nrows, ncols))
    println("Part 2: ", check_for_loops(boxes, guard, nrows, ncols))
end